src/remote_gym/remote_environment_management.py

```diff
@@ -222,64 +356,78 @@ def Process(self, request_iterator, context):
-                        action = unpacked_actions.get("action")
-                        observation, reward, terminated, truncated, info = self.environment.step(action)
+                    environment = self.get_environment(context.peer())
+
+                    unpacked_actions = environment.action_manager.unpack(internal_request.actions)
+                    action = unpacked_actions.get("action")
+
+                    observation, reward, terminated, truncated, info, rendering = environment.step(action)
 
-                    response = dm_env_rpc_pb2.StepResponse()
                     response_observations = {"observation": observation, "reward": reward}
 
-                    if self.rendering_enabled:
-                        rendering = self.environment.render()
+                    if self.enable_rendering:
                         response_observations.update({"rendering": rendering})
 
-                    packed_response_observations = observation_manager.pack(response_observations)
+                    packed_response_observations = environment.observation_manager.pack(response_observations)
+
+                    response = dm_env_rpc_pb2.StepResponse()
 
                     for requested_observation in internal_request.requested_observations:
                         response.observations[requested_observation].CopyFrom(
                             packed_response_observations[requested_observation]
                         )
                     if terminated or truncated:
                         response.state = dm_env_rpc_pb2.EnvironmentStateType.TERMINATED
-                        environment_should_be_reset = True
+                        environment.reset()
                     else:
                         response.state = dm_env_rpc_pb2.EnvironmentStateType.RUNNING
 
                 elif message_type == "reset":
-                    environment_should_be_reset = True
+                    environment = self.get_environment(context.peer())
+                    environment.reset()
+
                     response = dm_env_rpc_pb2.ResetResponse()
-                    for uid, action_space in self.action_spec.items():
+                    for uid, action_space in environment.action_spec.items():
                         response.specs.actions[uid].CopyFrom(action_space)
-                    for uid, observation_space in self.observation_spec.items():
+                    for uid, observation_space in environment.observation_spec.items():
                         response.specs.observations[uid].CopyFrom(observation_space)
 
                 elif message_type == "reset_world":
-                    environment_should_be_reset = True
+                    self.new_environment(context.peer(), self.enable_rendering)
+
                     response = dm_env_rpc_pb2.ResetWorldResponse()
 
                 elif message_type == "leave_world":
+                    self.destroy_environment(context.peer())
```

Comment:
I don't think you should do this when an agent leaves the world (just based on the protocol). In our case, an experiment ending should also send `message_type == "destroy_world"`, so no need to do it here additionally.